### PR TITLE
Ignore Exit status for Runnables in Google Batch

### PR DESCRIPTION
**Pull request by hnawar**
Ignore Exit status for runnables allows Errors to be ignored when specified in the workflow

Signed-off-by: Hatem Nawar <hnawar@google.com>

### CODE REVIEW

**Comment by hnawar** [comment]
#3166

**Comment by pditommaso** [comment]
Hi Hatem, thanks for looking into it. However, I'm not getting how this patch is related to #3166. 

If I'm understanding correctly `setIgnoreExitStatus(true)` applied to child jobs and makes them to not fail when the container task reports a non-zero exit status. 

Instead, the exception in #3166 is reported by the Nextflow runtime? how the ignore flag should affect it?

**Comment by aaronegolden** [comment]
Setting ignoreExitStatus(true) on the Runnable will cause Batch to ignore the exit status of that specific Runnable. If there are multiple Runnables in the Task then it means that Runnable can fail and Batch will just go along and execute the next Runnable. 

In the scenario from #3166, if I understand correctly, there is a Batch Job created for the "preseq (SRR389222_sub1)" NextFlow process, and that Job has a single Task, and that Task has a single Runnable. So if ignoreExitStatus is set on that Runnable then it really means the Job will never fail (unless something goes wrong so early in the process that the Job never runs at all, e.g. the VM can't be provisioned in the first place).

It sounds like what is actually desired for #3166 is for the Batch Job to be allowed to fail, but then for the NextFlow workflow to carry on in spite of that failure.

**Comment by hnawar** [comment]
Would refactoring to have 3 runnables (Staging, Main, Unstaging) with IgnoreExitStatus set to true for the main runnable be a better solution?

The bash script should capture the exit status of the user script/command and the unstaging runnable should still be able to run to copy partial output,logs, and other artifacts

**Comment by aaronegolden** [comment]
I think the Batch ignoreExitStatus field is a red herring. Life Sciences also has that field, and it's not used for the preseq process in #3166. This same process fails when using the Life Sciences backend, but the NextFlow workflow carries on because that part of the workflow uses errorStrategy "IGNORE".

https://github.com/nf-core/methylseq/blob/118077265374f8bd76f60472cc1ab21d116c2a1a/conf/base.config#L54
https://github.com/nf-core/methylseq/blob/master/modules/nf-core/preseq/lcextrap/main.nf#L4

Whether I run the methylseq workflow with executor google-lifesciences or executor google-batch, either way I see a message in my .nextflow.log saying that the preseq process failed, and that the error was ignored, like:

"NOTE: Process `preseq (SRR389222_sub2)` terminated with an error exit status (1) -- Error is ignored"

The difference is, when executor=google-lifesciences the error appears to *actually* be ignored, whereas when executor=google-batch the workflow terminates.

So I think that any changes to the Batch Job submitted by NextFlow for the preseq process, causing the Job to not fail, would just be workarounds. The Job should be allowed to fail, and NextFlow should still complete the workflow due to that process using errorStrategy = 'ignore', as it does when the equivalent Pipeline fails when using the google-lifesciences executor.

File: plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy [hunk 1]
```diff
@@ -165,7 +165,8 @@ class GoogleBatchTaskHandler extends TaskHandler {
             .setComputeResource(computeResource)
             .addRunnables(
                 Runnable.newBuilder()
-                    .setContainer(container)
+                        .setContainer(container)
+                        .setIgnoreExitStatus(true)
```
Comment (bentsherman) [inline plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy:169]:
```suggestion
                    .setContainer(container)
                    .setIgnoreExitStatus(true)
```

**Pull request closed without merging**